// @nightwatch/schematics/src/schematics/nightwatch/utility/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLatestNodeVersion = exports.removePackageJsonDependency = exports.parseJsonAtPath = exports.addPropertyToPackageJson = exports.readPackageJson = exports.getAngularVersion = void 0;
const core_1 = require("@angular-devkit/core");
const schematics_1 = require("@angular-devkit/schematics");
const https_1 = require("https");
const enums_1 = require("../enums");
const dependencies_1 = require("./dependencies");
const json_utils_1 = require("./json-utils");
function getAngularVersion(tree) {
    const packageNode = dependencies_1.getPackageJsonDependency(tree, '@angular/core');
    const version = packageNode &&
        packageNode.version
            .replace(/[~^]/, '')
            .split('.')
            .find((x) => !!parseInt(x, 10));
    return version ? +version : 0;
}
exports.getAngularVersion = getAngularVersion;
function readPackageJson(tree) {
    const buffer = tree.read(enums_1.Config.PackageJsonPath);
    if (buffer === null) {
        throw new schematics_1.SchematicsException('Could not read package.json.');
    }
    const content = buffer.toString();
    const packageJson = core_1.parseJsonAst(content, core_1.JsonParseMode.Strict);
    if (packageJson.kind != 'object') {
        throw new schematics_1.SchematicsException('Invalid package.json. Was expecting an object');
    }
    return packageJson;
}
exports.readPackageJson = readPackageJson;
function addPropertyToPackageJson(tree, context, propertyName, propertyValue) {
    const packageJsonAst = readPackageJson(tree);
    const pkgNode = json_utils_1.findPropertyInAstObject(packageJsonAst, propertyName);
    const recorder = tree.beginUpdate('package.json');
    if (!pkgNode) {
        // outer node missing, add key/value
        json_utils_1.appendPropertyInAstObject(recorder, packageJsonAst, propertyName, propertyValue, enums_1.Config.JsonIndentLevel);
    }
    else if (pkgNode.kind === 'object') {
        // property exists, update values
        for (let [key, value] of Object.entries(propertyValue)) {
            const innerNode = json_utils_1.findPropertyInAstObject(pkgNode, key);
            if (!innerNode) {
                // script not found, add it
                context.logger.debug(`creating ${key} with ${value}`);
                json_utils_1.insertPropertyInAstObjectInOrder(recorder, pkgNode, key, value, enums_1.Config.JsonIndentLevel);
            }
            else {
                // script found, overwrite value
                context.logger.debug(`overwriting ${key} with ${value}`);
                const { end, start } = innerNode;
                recorder.remove(start.offset, end.offset - start.offset);
                recorder.insertRight(start.offset, JSON.stringify(value));
            }
        }
    }
    tree.commitUpdate(recorder);
}
exports.addPropertyToPackageJson = addPropertyToPackageJson;
function parseJsonAtPath(tree, path) {
    const buffer = tree.read(path);
    if (buffer === null) {
        throw new schematics_1.SchematicsException('Could not read package.json.');
    }
    const content = buffer.toString();
    const json = core_1.parseJsonAst(content, core_1.JsonParseMode.CommentsAllowed);
    if (json.kind != 'object') {
        throw new schematics_1.SchematicsException('Invalid package.json. Was expecting an object');
    }
    return json;
}
exports.parseJsonAtPath = parseJsonAtPath;
// modified version from utility/dependencies
function removePackageJsonDependency(tree, dependency) {
    const packageJsonAst = parseJsonAtPath(tree, enums_1.pkgJson.Path);
    const depsNode = json_utils_1.findPropertyInAstObject(packageJsonAst, dependency.type);
    const recorder = tree.beginUpdate(enums_1.pkgJson.Path);
    if (!depsNode) {
        // Haven't found the dependencies key.
        new schematics_1.SchematicsException('Could not find the package.json dependency');
    }
    else if (depsNode.kind === 'object') {
        const fullPackageString = depsNode.text.split('\n').filter((pkg) => {
            return pkg.includes(`"${dependency.name}"`);
        })[0];
        const commaDangle = fullPackageString && fullPackageString.trim().slice(-1) === ',' ? 1 : 0;
        const packageAst = depsNode.properties.find((node) => {
            return node.key.value.toLowerCase() === dependency.name.toLowerCase();
        });
        // TODO: does this work for the last dependency?
        const newLineIndentation = 5;
        if (packageAst) {
            // Package found, remove it.
            const end = packageAst.end.offset + commaDangle;
            recorder.remove(packageAst.key.start.offset, end - packageAst.start.offset + newLineIndentation);
        }
    }
    tree.commitUpdate(recorder);
}
exports.removePackageJsonDependency = removePackageJsonDependency;
/**
 * Attempt to retrieve the latest package version from NPM
 * Return an optional "latest" version in case of error
 * @param packageName
 */
function getLatestNodeVersion(packageName) {
    const DEFAULT_VERSION = 'latest';
    return new Promise((resolve) => {
        return https_1.get(`https://registry.npmjs.org/${packageName}`, (res) => {
            let rawData = '';
            res.on('data', (chunk) => (rawData += chunk));
            res.on('end', () => {
                try {
                    const response = JSON.parse(rawData);
                    const version = (response && response['dist-tags']) || {};
                    if (packageName === 'nightwatch') {
                        resolve(buildPackage(packageName, version.next));
                    }
                    resolve(buildPackage(packageName, version.latest));
                }
                catch (e) {
                    resolve(buildPackage(packageName));
                }
            });
        }).on('error', () => resolve(buildPackage(packageName)));
    });
    function buildPackage(name, version = DEFAULT_VERSION) {
        return { name, version };
    }
}
exports.getLatestNodeVersion = getLatestNodeVersion;
